AGDIGGER-63

luigizuccarelli opened the pull request:
Update to create an s2i build for the  android sdk
    
## Motivation:
With the jenkins openshift platform the mounting of an external PV for the android sdk for each instance of a slave proved to be problamatic.
A new approach is to use a s2i image to include the installed sdk.
The thinking here was to allow the admin user to download install and accept the license agreement.
The s2i build would then include the sdk into a s2i slave base image. The resultant image would then be uploaded to the internal docker openshift registry.
    
## Changes:
A new directory structure was included into the repo , with all the valid scripts and Dockerfile to realize this functionality
    
## Goals:
The final goal is to use the slave base s2i image within an ansible script to automate various versions of the android sdk
    
JIRA:
Refer to https://issues.jboss.org/browse/AGDIGGER-63

luigizuccarelli commented:
@matzew @wtrocki Please review.

I have also created a PR for the upstream https://github.com/openshift/jenkins/pull/214

wtrocki commented on docker/README.md:
```diff
@@ -67,7 +67,14 @@ Jenkins Kubernetes plugin slaves:
 
 ## Build
```
It should be `### Build`

wtrocki commented on docker/README.md:
```diff
@@ -67,7 +67,14 @@ Jenkins Kubernetes plugin slaves:
 ## Build
 
-To build slave execute docker build
+S2I build is required in order to include the 'acpepted license agreement' android sdk
```
I would change this to: 
'acpepted license agreement' android sdk =>

We using s2i technology to for building mobile jenkins slaves. During s2i build specific version of the Android SDK is installed and developer is asked to confirm all required licenses.

wtrocki commented on docker/README.md:
```diff
@@ -67,7 +67,14 @@ Jenkins Kubernetes plugin slaves:
+    cd android-slave-s2i directory
+    docker build -t aerogear/jenkins-android-slave-s2i
 
+To include the android-sdk, download the required sdk, install and accept the license agreement
```
[Optional] I think that we may include here our example installation script. It's not easy to know what would be required sdk installation etc.

wtrocki commented on docker/android-slave-s2i/.s2i/bin/assemble:
```diff
@@ -0,0 +1,20 @@
+fi
+
+# Copy the android sdk to /opt/android-sdk-linux 
+# Also delete it from the /tmp/src directory avoiding docker bloat 
```
[Quesion] What is this docker bloat in this case?

wtrocki commented on docker/android-slave-s2i/.s2i/bin/usage:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash -e
```
👍

wtrocki commented on docker/android-slave-s2i/.s2i/bin/usage:
```diff
@@ -0,0 +1,12 @@
+s2i build <dir-where-accepted-adroid-sdk-is-installed> android-slave-s2i aerogear/jenkins-android-slave
+
+You can then run the resulting image via:
+docker run <application image>
```
replace `<application image>`  with aerogear/jenkins-android-slave

wtrocki commented on docker/android-slave-s2i/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+
```
Why another folder here? IMHO We should use existing folder and do not produce additional artifacts. 
It's hard to review this now because it's not easy to see what was changed. Having s2i capability doesn't change a lot here. I do not see changes here to justify new file.

wtrocki commented on docker/android-slave-s2i/contrib/bin/scl_enable:
```diff
@@ -0,0 +1,3 @@
+# This will make scl collection binaries work out of box.
```
Same as above. We shud not create another folder and use android-slave resources.
Let's avoid duplicates in the repo, which are hard to maintain and change.

wtrocki commented on docker/android-slave-s2i/test/run:
```diff
@@ -0,0 +1,160 @@
+#!/bin/bash
```
Is that from base-s2i image? Can we use image instead of copying file here? 
I'm asking because it would be easier to get that from upstream

wtrocki commented on docker/android-slave-s2i/test/run:
```diff
@@ -0,0 +1,160 @@
+  # s2i build requires the application is a valid 'Git' repository
+  pushd ${test_dir}/test-app >/dev/null
+  git init
+  git config user.email "build@localhost" && git config user.name "builder"
```
Why we doing it? It's really strange that we going to change this.

wtrocki commented on docker/android-slave-s2i/test/run:
```diff
@@ -0,0 +1,160 @@
+check_result $?
+
+# Verify that the HTTP connection can be established to test application container
+run_test_application &
```
This step is not clear for me and also all methods related with this.

wtrocki commented on docker/android-slave-s2i/test/run:
```diff
@@ -0,0 +1,160 @@
+      docker stop $(cat $cid_file)
+    fi
+  fi
+  if image_exists ${IMAGE_NAME}-testapp; then
```
Why `-testapp`?

wtrocki commented on docker/android-slave-s2i/tools/android-wait-for-emulator.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
We do not need this anymore. Correct me if I'm wrong? From where this is coming from?

luigizuccarelli replied:
@wtrocki - Yes makes sense I'm not sure anyone would want to use the crappy emulator

wtrocki commented on docker/jenkins1-centos/Dockerfile:
```diff
@@ -1,6 +1,14 @@
+RUN /usr/local/bin/plugins.sh /opt/openshift/configuration/plugins.txt
+
+# needed for gradlew 
+ENV JAVA_HOME=/etc/alternatives/java_sdk_1.8.0_openjdk
```
Do we need this? That sounds like something that should be in upstream image.

wtrocki commented on openshift/README.md:
```diff
@@ -52,6 +52,32 @@ Note that `mkdir` and `chmod` commands above should be executed in the Docker-ma
+Jenkins Configuration
+---------------------
+
+The android-sdk is not 'included in any docker container' and therefore a simple 'freestyle job needs to be created' that will contain the scripts to download and install the android sdk
```
That should be automated - I would show you how.

wtrocki commented on openshift/README.md:
```diff
@@ -52,6 +52,32 @@ Note that `mkdir` and `chmod` commands above should be executed in the Docker-ma
+
+1. Navigate to 'Add build step' and click on 'Execute shell'
+
+        Copy and paste the script found in jenkinsfiles/install-android-sdk.sh
```
That file is missing now. Is that actual ?

wtrocki replied:
This looks like some old file to me and needs to be reviewed entirely.

wtrocki commented on docker/android-slave/Dockerfile:
```diff
@@ -1,6 +1,13 @@
-FROM openshift/jenkins-slave-base-centos7
 
-MAINTAINER Wojciech Trocki <wtr@redhat.com>
+FROM aerogear/jenkins-slave-base:1.0.0
```
Move to first line.

wtrocki commented on docker/android-slave/Dockerfile:
```diff
@@ -1,6 +1,13 @@
+
+MAINTAINER Luigi Zuccarelli <lzuccare@redhat.com>
+
+ENV ANDROID_SLAVE_SDK_BUILDER=1.0.0
```
What is that? Having some versions here is anti-pattern.

wtrocki commented on docker/android-slave/Makefile:
```diff
@@ -0,0 +1,10 @@
+	docker build -t $(IMAGE_NAME) .
+
+.PHONY: test
+test:
```
That's not working for me. Best to remove it or provide different test.

wtrocki commented:
ping @aliok - just to get familiar with idea and changes we doing!

aliok commented:
will review as soon as I can!

luigizuccarelli commented:
@wtrocki - I have updated the PR with a run script. I was chatting with Ben Parees , the s2i image **must** have a build script. I also added a BUILD_S2I envar - this ensures the build assemble script will be called and not the original entry point /usr/local/bin/run-jnlp-client. I have closed the upstream PR as it is not needed now. I still need to do an end-to-end test though

wtrocki commented:
@luigizuccarelli - Is upstream team interested in s2i capability?

In relation to your issues with checking git source code - that's bug.
You should use https url initially because ssh setup is not working properly. 
I going to log ticket for that. Sorry for delay in this - I just did not realized that this was a problem because I used http in my jobs.

luigizuccarelli commented:
@wtrocki - Ok I will use https to test it out. Ben made no mention or showed any interest so I'm not going to push it

wtrocki commented on docker/README.md:
```diff
@@ -65,9 +65,27 @@ Jenkins Kubernetes plugin slaves:
+    docker build -t aerogear/jenkins-android-slave-s2i
+
+To include the android-sdk, download the required sdk, install and accept the license agreement
+The following script is an example of downloading the sdk installing and lastly a step to accept the license agreement
```
replace `downloading and installing SDK`.  remove `step to accept the license agreement` - it's just part of installation.

wtrocki commented on docker/README.md:
```diff
@@ -65,9 +65,27 @@ Jenkins Kubernetes plugin slaves:
+    wget --output-document=android-sdk.tgz --quiet https://dl.google.com/android/android-sdk_r24.4.1-linux.tgz
+    tar xzf android-sdk.tgz 
+    
+    android-sdk-linux/tools/android update sdk --all --no-ui --filter platform-tools,tools,build-tools-25.0.0,android-25,addon-google_apis_x86-google-21,extra-android-support,extra-google-google_play_services,sys-img-armeabi-v7a-android-24
```
No need for emulator image for the moment - it would have significant amount of space (around 400mb) - `sys-img-armeabi-v7a-android-24`

wtrocki commented on docker/README.md:
```diff
@@ -65,9 +65,27 @@ Jenkins Kubernetes plugin slaves:
+
+    s2i build <directory-where-sdk-has-been-installed> aerogear/jenkins-android-slave-s2i aerogear/jenkins-android-slave:<version>
+
+For release builds, publish the image to the openshift internal registry
```
That may sound cryptic. Let's put some documentation from openshift here.

[Question] Can we leverage openshift to build s2i images instead of users doing them manually. Then this would be published automatically. Obvious drawback here is that we would need to figure out how to feed SDK files

luigizuccarelli replied:
@wtrocki - great idea - i.e everything happens in the openshift env.

wtrocki commented on docker/android-slave/Dockerfile:
```diff
@@ -24,30 +28,31 @@ RUN yum install -y centos-release-scl-rh && \
+    chmod -R g+rw $HOME
 
-# Install build dependencies
-    yum install -y git wget python curl expect expectk
```
How this works without git? Is git available in this image?

luigizuccarelli replied:
@wtrocki - Yes previously installed

wtrocki commented on docker/android-slave/Dockerfile:
```diff
@@ -24,30 +28,31 @@ RUN yum install -y centos-release-scl-rh && \
+# This default user is created in the openshift/base-centos7 image
 USER 1001
+WORKDIR /opt/workspace
+CMD ["usage"]
```
👍

luigizuccarelli commented on docker/android-slave/Dockerfile:
```diff
@@ -24,30 +28,31 @@ RUN yum install -y centos-release-scl-rh && \
-    rm -f android-sdk.tgz && \
-    chmod -R 777 /opt/tools
+# Setup directory for the android sdk 
+RUN mkdir -p /opt/android-sdk-linux && chmod -R 777 /opt/android-sdk-linux
```
@wtrocki - so just do mkdir and leave out the chmod ?

wtrocki replied:
That's created as root so we just need to change owner?

wtrocki commented on docker/android-slave/.s2i/bin/run:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -e
```
We need to document that this is now main entry point. Basically little bit documentation why this is used to run .  I know that's obvious for someone who knows s2i stuff, but calling this as main entrypoint for image with source it's not standard approach and it would be good to expose why we have it. Especially:
`/usr/local/bin/run-jnlp-client "$@"`

wtrocki replied:
@luigizuccarelli I meant to document this here - Just add additional comment here

luigizuccarelli commented:
@wtrocki - Ok update and rebased - one final look please

wtrocki commented:
@luigizuccarelli let's publish image to dockerhub using different label. For example: `sdk24-sti` then we can verify that on public cluster and change references in default kubernetes pod configuration as part of this ticket. Rebase(squash/rename) recommended for some commits.

wtrocki commented on docker/README.md:
````diff
@@ -63,9 +63,60 @@ Jenkins Kubernetes plugin slaves:
+The script (see below) will first check the params that are passed to it when executing the actual `s2i build` command, if the requirements are met it will follow the normal build flow.
+When called by the slave with parameters the original entry point script will be called.
+
+```
````
-1 for duplication. Can we reference file here. Let's keep things simple and.. short.

luigizuccarelli commented:
@wtrocki @aliok - I will merge and close once I have done sync meeting with Ali tomorrow

wtrocki commented:
@luigizuccarelli - we done quite productive review here and I'm really confident that logic here is really solid. Going to merge it to avoid long running PR and merge conflicts. We can extend this in different PR's. I'm going to do some standard cleanup in this repo and it would be nice to include this as well.
